File: c4_linkedlist/L23-h.cpp
//
// Created by 李恒 on 2020/1/8.
//

#include <iostream>
#include <vector>
using namespace std;

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size() == NULL)
            return NULL;
        ListNode* res = lists[0];
        for(int i=1 ; i< lists.size();i++)
            res = merge2Lists(res , lists[i]);
        return res;
    }

    ListNode* merge2Lists(ListNode* pHead1 , ListNode* pHead2 ){
        if(pHead1 == NULL)
            return pHead2;
        if(pHead2 == NULL)
            return pHead1;
        ListNode* new_head = new ListNode(0);
        ListNode* tem = new_head;

        while(pHead1 != NULL && pHead2 != NULL){
            if(pHead1->val > pHead2->val){
                tem->next = pHead2;
                pHead2 = pHead2->next;
            }
            else{
                tem->next = pHead1;
                pHead1 = pHead1->next;
            }
            tem = tem->next;
        }
        tem->next = (pHead1 == NULL)? pHead2:pHead1 ;
        ListNode* re = new_head->next ;
        delete new_head;
        return re;
    }

};



//采用的是两两合并链表，借助一个队列，直到队列里面元素只剩1个
#include <queue>
class Solution1 {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size() == 0)
            return NULL;
        if(lists.size() == 1)
            return lists[0];
        queue<ListNode*> q1;
        for(int i=0 ;i < lists.size();i++)
            q1.push(lists[i]);
        while(q1.size()>1){
            ListNode* l1 = q1.front();
            q1.pop();
            ListNode* l2 = q1.front();
            q1.pop();
            ListNode* p = merge2Lists(l1,l2);
            q1.push(p);
        }
        return q1.front();
    }

    ListNode* merge2Lists(ListNode* pHead1, ListNode* pHead2){
        if(pHead1 == NULL)
            return pHead2;
        if(pHead2 == NULL)
            return pHead1;
        ListNode* newHead = new ListNode(0);
        ListNode* tem = newHead;
        while(pHead1 && pHead2){
            if(pHead1->val > pHead2->val){
                tem->next = pHead2;
                pHead2 = pHead2->next;
            }
            else{
                tem->next = pHead1;
                pHead1 = pHead1->next;
            }
            tem = tem->next;
        }
        tem->next = pHead1 ? pHead1:pHead2;
        ListNode* res = newHead->next;
        delete newHead;
        return res;
    }

};

File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L23-h.py
#--coding:utf-8--
'''
@Time   : 2020/8/14
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# 使用数组模拟队列，两两合并
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if lists == []:
            return None
        que = []
        for lk in lists:
            que.append(lk)
        while len(que) > 1:
            l1 = que.pop(0)
            l2 = que.pop(0)
            tmp = self.merge2Lists(l1, l2)
            que.append(tmp)
        return que[0]

    def merge2Lists(self, l1, l2):
        dummy = ListNode(-1)
        cur = dummy
        while l1 and l2 :
            if l1.val < l2.val:
                cur.next = l1
                l1 = l1.next
            else:
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        if l1:
            cur.next = l1
        else:
            cur.next = l2
        return dummy.next